repos/backend/endpoints/files.js

```diff
@@ -43,7 +45,97 @@ const deleteFile = (app, config) => async (req, res) => {
   }
 }
 
+/**
+ * Recursively checks to find the parent node for a given item
+ * if a parent exists, it will add it as a child
+ * @param {Array<Object>} nodes 
+ * @param {string} parentPath 
+ * @param {Object} newItem
+ * 
+ * @returns {Boolean} - whether a parent node exists and push was successful
+ */
+const parentNodeExists = (nodes, parentPath, newItem) => {
+  
+  const found = nodes.find((node) => {
+    if (node.fullPath === parentPath) {
+      node.children.push(newItem)
+      return true
+    } 
+    else {
+      // check children
+      return node.children.length > 0 && parentNodeExists(node.children, parentPath, newItem)
+    }
+  })
+
+  return Boolean(found)
+}
+
+/**
+ * Transforms the paths string to a specific data object
+ * @param {Array<string>} paths - full paths to the folder or file i.e '/keg/tap/tests/bdd/features'
+ * 
+ * @returns {Array<Object>} - each object has the form: 
+ *                            {id, fullPath, children: [], isModified}
+ */
+const generateTree = (paths) => {
```

Comment:
See notes at the end for more explanation. 
I would update this to look more like this => 
```js
/**
 * Gets the metadata of a path from the local filesystem
 * @param {string} paths - full path to the folder or file i.e '/keg/tap/tests/bdd/features'
 * 
 * @returns {Object} - Meta data containing {name, parent, type ( folder || file )} properties
 */
const getPathMeta = path => {
  const pathSplit = path.split('/')
  return {
    name: pathSplit.pop(),
    parent: pathSplit.join('/'),
    type: fs.lstatSync(path).isDirectory() ? 'folder' : 'file',
  }
}

/**
 * Transforms the paths string to a specific data object
 * @param {Array<string>} paths - full paths to the folder or file i.e '/keg/tap/tests/bdd/features'
 * 
 * @returns {Array<Object>} - Each object contains {id, fullPath, children: [], isModified} properties
 */
const generateTree = (paths) => {
  /**
   * 1. create new object for each 'path' item
   * 2. if the parent path of current 'path' item exists, add it as the child
   */
  return paths.reduce((nodes, path) => {
    // Get the meta data for this path
    const { parent, ...pathMeta } = getPathMeta(path)
    
      // Ignore hidden files that start with a .
    if (pathMeta.type === 'file' && pathMeta.name.startsWith('.')) return nodes
  
    const node = {
      id: path,
      children: [],
      fullPath: path,
      isModified: false,
      ...pathMeta,
    }

    ;(!nodes.length || !parentNodeExists(nodes, parent, node)) &&
      nodes.push(node)

    return nodes
  }, [])

}
```
Notes:
* Don't create the node object until after we ensure it's actually needed... i.e. It's not a hidden file
* Extract the path meta data to it's own method, for read ability, and to clean things up
* Nothing wrong with `path.lastIndexOf` and `path.substring`, it just feels overly verbose
  * It can be done more simply in my opinion with `split` and `pop`
*  When checking `array.length`, the use of `> 0` is redundant.
  *  Because `0` is considered `falsy`, you can just check for the length => `array.length`
  * If it's `0`, then it will be `false`, otherwise it will be `true`
* Because `array.push` returns the length of the array after the new item has been added, it will always be true
  * Which means we can do this =>  `Boolean(array.push(item))` === `true`

Reply:
makes sense, this is much cleaner. thanks for the detailed notes
updated